webap/seechange_webap.py

```diff
@@ -0,0 +1,478 @@
+
+@app.route( "/", strict_slashes=False )
+def mainpage():
+    return flask.render_template( "seechange_webap.html" )
+
+# **********************************************************************
+
+@app.route( "/exposures", methods=['POST'], strict_slashes=False )
+def exposures():
+    try:
+        data = { 'startdate': None,
+                 'enddate': None
+                }
+        if flask.request.is_json:
+            data.update( flask.request.json )
+
+        app.logger.debug( f"After parsing, data = {data}" )
+        t0 = None if data['startdate'] is None else astropy.time.Time( data['startdate'], format='isot' ).mjd
+        t1 = None if data['enddate'] is None else astropy.time.Time( data['enddate'], format='isot' ).mjd
+        app.logger.debug( f"t0 = {t0}, t1 = {t1}" )
+
+        conn = next( dbconn() )
+        cursor = conn.cursor()
+        # TODO : deal with provenance!
+        # (We need some kind of provenance tagging table, so that the user can specify
+        # a user-readable name (e.g. "default", "latest", "dr1", whatever) that specifies
+        # a set of provenances to search.  One of these names must be all the provenances
+        # we're using "right now" in the active pipeline; that will be the one that
+        # (by default) the webap uses.
+        q = ( 'SELECT m.id, m.filepath, m.mjd, m.target, m.filter, m.filter_array, m.exp_time, '
+              '       m.n_images, m.n_cutouts, m.n_measurements, '
+              '       SUM( CASE WHEN r.success THEN 1 ELSE 0 END ) AS n_successim, '
+              '       SUM( CASE WHEN r.error_message IS NOT NULL THEN 1 ELSE 0 END ) as n_errors '
+              'FROM ( '
+              '  SELECT e.id, e.filepath, e.mjd, e.target, e.filter, e.filter_array, e.exp_time, '
+              '         COUNT(DISTINCT(i.id)) AS n_images, COUNT(c.id) AS n_cutouts, COUNT(m.id) AS n_measurements '
+              '  FROM exposures e '
+              '  LEFT JOIN images i ON i.exposure_id=e.id '
+              '  LEFT JOIN image_upstreams_association ias ON ias.upstream_id=i.id '
+              '  LEFT JOIN images s ON s.id = ias.downstream_id AND s.is_sub '
+              '  LEFT JOIN source_lists sl ON sl.image_id=s.id '
+              '  LEFT JOIN cutouts c ON c.sources_id=sl.id '
+              '  LEFT JOIN measurements m ON m.cutouts_id=c.id '
+              '  LEFT JOIN reports r ON r.exposure_id=e.id ' )
+        subdict = {}
+        if ( t0 is not None ) or ( t1 is not None ):
+            q += "  WHERE "
+            if t0 is not None:
+                q += 'e.mjd >= %(t0)s'
+                subdict['t0'] = t0
+            if t1 is not None:
+                if t0 is not None: q += ' AND '
+                q += 'e.mjd <= %(t1)s'
+                subdict['t1'] = t1
+        q += ( '   GROUP BY e.id ' # ,e.filepath,e.mjd,e.target,e.filter,e.filter_array,e.exp_time '
+               '   ORDER BY e.mjd, e.filter, e.filter_array ' )
+
```

Comment:
...so this is what it looks like to work without an ORM...?

Reply:
I wouldn't even want to think about figuring out the sqlalchemy idiom that would correspond to this query.  It would probably take a me a week to figure out in the first place, and another 3 weeks to debug.

Reply:
I think it would be something like: all the joins are done automatically using relationships, and all the sums you will just do on the python side. Less efficient but probably easier to write.

Reply:
...well, at least we can agree that it would be less efficient.